Source/WebCore/platform/text/SegmentedString.cpp

```diff
@@ -182,18 +182,19 @@ void SegmentedString::advanceAndUpdateLineNumber16()
 inline void SegmentedString::advancePastSingleCharacterSubstringWithoutUpdatingLineNumber()
 {
     ASSERT(m_currentSubstring.length == 1);
+
+    m_currentSubstring.length = 0;
+    m_numberOfCharactersConsumedPriorToCurrentSubstring += m_currentSubstring.numberOfCharactersConsumed();
     if (m_otherSubstrings.isEmpty()) {
-        m_currentSubstring.length = 0;
+        m_currentSubstring = { };
         m_currentCharacter = 0;
```

Comment:
This is a case that we’d like to optimize. Merging the code with the normal case may be aesthetically appealing and does fix one specific problem here, but we may be forgoing some possibly-valuable efficiency in what should be the common case by using `updateAdvanceFunctionPointers` instead of `updateAdvanceFunctionPointersForEmptyString`, and even by repeatedly zeroing out `m_currentSubstring`.

I’d like to see more rationale for exactly why this is the best change. Despite the bug we are fixing here, I liked the old version with the early return better and this doesn’t seem like a clear improvement.

Reply:
@darinadler You're right about `updateAdvanceFunctionPointersForEmptyString()` and early return. I missed that these are two different methods. But we have to reset `m_currentSubstring`. If we don't do that `SegmentedString::numberOfCharactersConsumed()` will give a wrong result later. That was the reason for new test failures.